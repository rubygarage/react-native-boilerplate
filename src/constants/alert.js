// eslint-disable-next-line import/prefer-default-export
export const ALERT_TYPES = {
  error: 'error',
  success: 'success',
  info: 'info',
};
