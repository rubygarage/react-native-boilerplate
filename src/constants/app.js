export const GLOBAL_ERROR = '_error';
export const OFFLINE_FLASH_ID = 'offline';
export const OFFLINE_FLASH_PRIORITY = 0;
export const FLASH_LIFETIME = 3000;

export const PLATFORM_TYPE = {
  android: 'android',
  ios: 'ios',
};
