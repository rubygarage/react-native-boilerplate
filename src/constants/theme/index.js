export default {
  colors: {
    black: '#000000',
    white: '#ffffff',
  },
  fonts: {
  },
  sizes: {
  },
};
