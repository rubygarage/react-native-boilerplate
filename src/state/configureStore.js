import { createStore, applyMiddleware } from 'redux';
import { createLogicMiddleware } from 'redux-logic';
import { composeWithDevTools } from 'redux-devtools-extension/developmentOnly';
import { persistStore, persistReducer } from 'redux-persist';
import { Settings } from 'luxon';
import AsyncStorage from '@react-native-community/async-storage';

import buildHttpClient from 'lib/httpClient';
import { rootHttpClient } from 'lib/httpClient/buildHttpClient';
import mountInterceptors from 'lib/httpClient/mountInterceptors';
import conceptsOperations from './concepts/operationsRoot';
import flashOperations from './flash/operations';
import rootReducer from './store';

const operations = [...conceptsOperations, ...flashOperations];

const configureStore = () => {
  const persistConfig = {
    key: 'root',
    storage: AsyncStorage,
    whitelist: ['data', 'application'],
    timeout: null,
  };
  const persistedReducer = persistReducer(persistConfig, rootReducer);

  const operationsDependencies = {
    httpClient: buildHttpClient(),
    rootHttpClient,
  };

  const logicMiddleware = createLogicMiddleware(operations, operationsDependencies);
  const middlewares = applyMiddleware(logicMiddleware);
  const enhancer = composeWithDevTools(middlewares);

  const store = createStore(persistedReducer, enhancer);
  const persistor = persistStore(store);
  store.logicMiddleware = logicMiddleware;
  store.httpClient = operationsDependencies.httpClient;

  mountInterceptors(store);

  if (module.hot) {
    module.hot.accept(() => {
      // eslint-disable-next-line global-require
      store.replaceReducer(require('./store').default);
    });
  }

  Settings.defaultLocale = store.getState().intl.locale;

  return { store, persistor };
};

export default configureStore;
