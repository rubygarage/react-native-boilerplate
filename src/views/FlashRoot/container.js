import React, { Component } from 'react';
import { connect } from 'react-redux';
import { UIManager, LayoutAnimation } from 'react-native';
import PropTypes from 'prop-types';

import { flashesSelector } from 'state/flash/selectors';
import FlashRootComponent from './component';

const ANIMATION_DURATION = 250;

class FlashRoot extends Component {
  componentDidMount() {
    if (UIManager.setLayoutAnimationEnabledExperimental) {
      UIManager.setLayoutAnimationEnabledExperimental(true);
    }
  }

  componentDidUpdate(prevProps) {
    const { flashes: prevFlashes } = prevProps;
    const { flashes } = this.props;
    if (prevFlashes.length < flashes.length) {
      this.doAnimation();
    }
  }

  doAnimation = () => {
    LayoutAnimation.configureNext({
      duration: ANIMATION_DURATION,
      update: {
        type: LayoutAnimation.Types.linear,
      },
    });
    LayoutAnimation.configureNext({
      duration: ANIMATION_DURATION,
      create: {
        type: LayoutAnimation.Types.linear,
        property: LayoutAnimation.Properties.scaleXY,
      },
    });
  };

  render() {
    return (
      <FlashRootComponent
        {...this.props}
        doAnimation={this.doAnimation}
      />
    );
  }
}

FlashRoot.propTypes = {
  flashes: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
};

const mapStateToProps = (state) => ({
  flashes: flashesSelector(state),
});

export { FlashRoot as FlashRootContainer };
export default connect(mapStateToProps)(FlashRoot);
