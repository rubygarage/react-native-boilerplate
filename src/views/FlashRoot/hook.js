import { useEffect } from 'react';
import { useSelector } from 'react-redux';
import { UIManager, LayoutAnimation } from 'react-native';

import { flashesSelector } from 'state/flash/selectors';

export const ANIMATION_DURATION = 250;

const useContainer = () => {
  const flashes = useSelector(flashesSelector);

  const doAnimation = () => {
    LayoutAnimation.configureNext({
      duration: ANIMATION_DURATION,
      update: {
        type: LayoutAnimation.Types.linear,
      },
    });
    LayoutAnimation.configureNext({
      duration: ANIMATION_DURATION,
      create: {
        type: LayoutAnimation.Types.linear,
        property: LayoutAnimation.Properties.scaleXY,
      },
    });
  };

  useEffect(() => {
    if (UIManager.setLayoutAnimationEnabledExperimental) {
      UIManager.setLayoutAnimationEnabledExperimental(true);
    }
  }, []);

  useEffect(doAnimation, [flashes.length]);

  return {
    doAnimation,
  };
};

export default useContainer;
