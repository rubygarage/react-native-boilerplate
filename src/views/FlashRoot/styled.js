import styled from 'styled-components/native';

const List = styled.FlatList`
  position: absolute;
  top: ${({ topInset }) => topInset};
  width: 100%;
  margin-top: 12;
  margin-bottom: 12;
  padding-left: 12;
  padding-right: 12;
  z-index: 5;
`;

// eslint-disable-next-line import/prefer-default-export
export { List };
