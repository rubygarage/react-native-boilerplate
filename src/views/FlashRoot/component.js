import React from 'react';
import PropTypes from 'prop-types';

import Flash from './Flash';
import * as S from './styled';

const keyExtractor = (item) => item.id;

const FlashRootComponent = ({ flashes, doAnimation, topInset }) => (
  <S.List
    data={flashes}
    renderItem={({ item }) => <Flash {...item} doAnimation={doAnimation} />}
    keyExtractor={keyExtractor}
    showsVerticalScrollIndicator={false}
    topInset={topInset}
  />
);

FlashRootComponent.propTypes = {
  flashes: PropTypes.arrayOf(PropTypes.shape({})).isRequired,
  doAnimation: PropTypes.func.isRequired,
  topInset: PropTypes.number,
};

FlashRootComponent.defaultProps = {
  topInset: 0,
};

export default FlashRootComponent;
