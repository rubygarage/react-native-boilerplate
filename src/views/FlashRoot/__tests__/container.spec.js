import React from 'react';
import { shallow } from 'enzyme';
import configureStore from 'redux-mock-store';
import { UIManager } from 'react-native';

import diveTo from 'utils/testHelpers/diveToEnzyme';
import { ALERT_TYPES } from 'constants/alert';
import FlashRoot, { FlashRootContainer } from '../container';

jest.mock('react-native/Libraries/ReactNative/UIManager', () => ({
  getViewManagerConfig: jest.fn(),
}));

jest.mock('react-native/Libraries/LayoutAnimation/LayoutAnimation', () => ({
  ...jest.requireActual(
    'react-native/Libraries/LayoutAnimation/LayoutAnimation',
  ),
  configureNext: jest.fn(),
}));

jest.mock('state/flash/selectors', () => ({
  flashesSelector: jest.fn(() => []),
}));

describe('FlashRoot container', () => {
  const store = configureStore()({});
  store.dispatch = jest.fn();
  const defaultProps = { store };
  const wrapper = shallow(<FlashRoot {...defaultProps} />);
  const container = diveTo(wrapper, FlashRootContainer);
  const instance = container.instance();

  beforeEach(() => {
    jest.clearAllMocks();
  });

  it('renders correctly', () => {
    expect(container).toMatchSnapshot();
  });

  it('componentDidMount', () => {
    const mockSetLayoutAnimationEnabledExperimental = jest.fn();
    UIManager.setLayoutAnimationEnabledExperimental = mockSetLayoutAnimationEnabledExperimental;
    instance.componentDidMount();
    expect(mockSetLayoutAnimationEnabledExperimental).toHaveBeenCalledWith(true);
  });

  it('componentDidUpdate calls "doAnimation"', () => {
    const doAnimationSpy = jest.spyOn(instance, 'doAnimation');
    container.setProps({ flashes: [{ id: 0, type: ALERT_TYPES.info }] });
    instance.componentDidUpdate({ flashes: [] });

    expect(doAnimationSpy).toHaveBeenCalled();
  });

  it('componentDidUpdate does not call "doAnimation"', () => {
    const doAnimationSpy = jest.spyOn(instance, 'doAnimation');
    container.setProps({ flashes: [] });
    doAnimationSpy.mockClear();
    instance.componentDidUpdate({ flashes: [] });

    expect(doAnimationSpy).not.toHaveBeenCalled();
  });
});
