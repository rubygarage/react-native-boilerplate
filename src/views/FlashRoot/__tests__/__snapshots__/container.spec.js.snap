// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`FlashRoot container renders correctly 1`] = `
<FlashRootComponent
  dispatch={[MockFunction]}
  doAnimation={[Function]}
  flashes={Array []}
  store={
    Object {
      "clearActions": [Function],
      "dispatch": [MockFunction],
      "getActions": [Function],
      "getState": [Function],
      "replaceReducer": [Function],
      "subscribe": [Function],
    }
  }
  topInset={0}
/>
`;
